feat: Migrate S3 Client from AWS SDK v2 (deprecated) to v3

vahidalizad opened the pull request:
## Summary
This pull request migrates the S3 client implementation from AWS SDK v2 to v3.

## Changes
1. **Initialization**: Updated the S3 client initialization to use `S3Client` from `@aws-sdk/client-s3` package.
2. **Commands**: Replaced v2 commands (`getObject`, `putObject`, etc.) with their v3 equivalents (`GetObjectCommand`, `PutObjectCommand`, etc.).
3. **Response Handling**: Modified response handling to accommodate changes in the v3 API, particularly in how streams and buffers are managed.
5. **Testing**: Added unit tests to cover new functionality and ensure compatibility with the existing codebase. Implemented mocks for offline testing using Jasmine.

## Release Dependency
This pull request is intended for after the release of the asynchronous `getFileLocation` function on the `parse-server` repository. Please ensure that this dependency is met before merging.

The relevant pull request for adding support of asynchronous `getFileLocation` in adapters can be found here: [parse-server PR #9271](https://github.com/parse-community/parse-server/pull/9271).

## References
- [AWS SDK for JavaScript v3 Documentation](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/)

Please review the changes and provide feedback. Thank you for your time and consideration.

Closes: #197

mtrezza commented:
Continuing from https://github.com/parse-community/parse-server-s3-adapter/pull/220#issuecomment-2395340067.

vahidalizad commented:
> Continuing from [#220 (comment)](https://github.com/parse-community/parse-server-s3-adapter/pull/220#issuecomment-2395340067).
> 
> I notice that this CI does not test against Parse Server, so with or without [parse-community/parse-server#9271](https://github.com/parse-community/parse-server/pull/9271), the CI passes. That is not good, because w/o integration test we cannot really know whether this all works together. That has been less of an issue in the past, where this adapter just had to comply with the interface. But now that it depends on a specific implementation of the interface (async support for interface methods), it would be good to test. Since none of the parse server adapters does integration tests, this is not an exception, and we could go ahead and merge this, but someone should test this end-to-end and confirm this is working before we merge.

I completely agree that testing integrations is crucial, especially given the dependency on async support for interface methods. That said, I think we could merge this into a separate branch for now and make adding integration tests a separate task. We could raise an issue for that and have others contribute to the testing. Once the tests are complete and we’ve confirmed everything works end-to-end, we can merge it into the master branch.

mtrezza commented:
Our CD automation setup currently does not allow maintaining feature branches. Hence we'll have to merge this into master. 

I think it would be enough if you could test this out with the latest version of Parse Server that supports the async interface. Maybe @mman also has time to test this out, just to confirm it works. Then we can go ahead and merge this and open a Server e2e test as a separate issue for later on.

mman commented:
@mtrezza I would love to test it but currently I am low-priority stuck on two conflicting issues:

```
const ParsePushAdapter = require('@parse/push-adapter').ParsePushAdapter;
```

that I used to work around push adapter version issues fails with: `Error [ERR_REQUIRE_ESM]: require() of ES Module not supported.`. 

And when I convert my code to ESM, parse server fails to start with (simplified):

```
Error [ERR_REQUIRE_ESM]: require() of ES Module ./cloud/main.js from ./node_modules/parse-server/lib/ParseServer.js not supported.
```

not sure what is the solution yet but will have a time to look into it next week...

mtrezza commented:
@mman does this help you? https://github.com/parse-community/parse-server/issues/9316#issuecomment-2379043778

mman commented:
> @mman does this help you? [parse-community/parse-server#9316 (comment)](https://github.com/parse-community/parse-server/issues/9316#issuecomment-2379043778)

I think it's actually this one: https://github.com/parse-community/parse-server/issues/7559 with workaround mentioned here: https://stackoverflow.com/questions/77638089/how-to-make-parse-server-detect-im-using-esm-now where instead of passing a cloud code path to Parse Server constructor, you actually have to `import` cloud code in your main .js file and pass the imported code to the constructor...

vahidalizad commented:
OK in that case I'm gonna add the new integration tests too and try to sync the dependencies with the parse-server itself lets have a cleaning task as well because all the codes have changed since the last update

mtrezza commented:
That would be great, maybe we should do that in a different PR though and keep this one open. Just because we now have a presumably working PR that we could merge in case the integration test PR stales.

mtrezza commented:
@vahidalizad Could you please revert all the changes that are unrelated? I believe we should leave this PR in the state of https://github.com/parse-community/parse-server-s3-adapter/pull/221/commits/d4062cbc78a5514a7ee0ead554a9094fd093f555.

vahidalizad commented:
> @vahidalizad Great to see you being so quick to add the changes :-) However, could you please revert the changes that are unrelated? I believe we should leave this PR in the state of [d4062cb](https://github.com/parse-community/parse-server-s3-adapter/commit/d4062cbc78a5514a7ee0ead554a9094fd093f555).
> 
> The changes you made in the last 2 commits may have side effects. I see for example some dependency upgrades that cannot be done like that will make workflows fail. That's why we always have separate PRs for separate issues, see also our [Contribution Policy, section Scope](https://github.com/parse-community/parse-server/blob/alpha/CONTRIBUTING.md#scope).

I apologize for any confusion. My initial intention was to refactor the entire codebase to reduce technical debt. However, I've reverted the dependency changes and only updated Jasmine and NYC. If that poses any issues, I can revert those as well.

I also implemented integration tests with the latest version of Parse Server. Due to my new environment being on Windows, I was unable to run the `test:integration` command directly. Instead, I used Docker to run my MongoDB instance rather than using `mongodb-runner`. If you have the opportunity, it would be great if you could take a look at that.

mtrezza commented:
> only updated Jasmine and NYC.

If it's not related to this PR, then it would be better to revert. I appreciate your intention, but unfortunately we have seen in the past that these combined changes of unrelated stuff can sometimes consume a lot of time when there are issues. Also, if there is an issue with these, we'd have to revert the whole PR, including the actual feature.

vahidalizad commented:
Thank you for the feedback! This makes total sense, and I understand this is the standard practice. The reason I combined the changes in a single PR was to avoid dealing with the merge conflict that would arise from writing integration tests for the old `makeS3Adapter` function (from the v2 version) and later resolving conflicts once this branch with the new `makeS3Adapter` function is merged.

That said, I understand the importance of keeping PRs focused, so I’ll follow your suggestion and create a separate PR for the integration tests.

Regarding the nyc upgrade, I think we should update this in a separate PR as well, because parse-server uses nyc in all other cases, and the current version of istanbul shows a high vulnerability in the npm audit.

mtrezza commented:
>  with the merge conflict that would arise from writing integration tests for the old makeS3Adapter function (from the v2 version) and later resolving conflicts once this branch with the new makeS3Adapter function is merged.

You are right, there is no need to write extensive integration tests for code that would be discarded anyway. But I assume there is basic adapter functionality that will not change, regardless of which AWS SDK is used. For if you could just write 1 integration test for that, that's all we need. It's just to show that the integration setup works, to move it into a separate PR.

vahidalizad commented:
> > with the merge conflict that would arise from writing integration tests for the old makeS3Adapter function (from the v2 version) and later resolving conflicts once this branch with the new makeS3Adapter function is merged.
> 
> You are right, there is no need to write extensive integration tests for code that would be discarded anyway. But I assume there is basic adapter functionality that will not change, regardless of which AWS SDK is used. If you could just write 1 integration test for that, that's all we need. It's just to show that the integration setup works, to move it into a separate PR. I assume that the integration test can be added without any changes in the adapter code itself, as it's an addition rather than a modification, so there really shouldn't be any merge conflict.

Thanks for the clarification! To answer your point, there is a mock adapter that changes between v2 and v3, which is used in both unit and integration tests. I was trying to avoid conflicts related to that since it's in the tests and not the main code, but it doesn’t matter anymore.

I’ve created a new PR (#224) that adds the integration tests as requested. Please review and merge it when possible. After that, I’ll revert this PR to the state before adding integration tests, resolve the conflicts, and commit the updates so this PR can be finalized after we merge #224.

mtrezza commented:
@vahidalizad I hope resolving the conflicts won't be too much work

vahidalizad commented:
> @vahidalizad I hope resolving the conflicts won't be too much work, but great that we have the integration tests now

I believe I’ve resolved all the conflicts and pushed the changes, but due to the latest updates, I need to create an S3 bucket and test it locally just to ensure nothing has changed there. I wrote a mock S3 adapter, but double-checking won’t hurt, so please hold off on merging for now. Also, there seems to be an issue with Parse Server 6, which I think might be related to the async function we modified.

vahidalizad commented:
I've tested this adapter with a real S3 bucket, and it worked flawlessly. I believe it's ready for migration. However, before proceeding, please review the code and feel free to clean it up if you'd like to create a more standardized adapter template.

mtrezza commented on spec/test.spec.js:
```diff
@@ -1,8 +1,11 @@
-const AWS = require('aws-sdk');
+const { Readable } = require('stream');
+const Parse = require('parse').Parse;
```
`Parse` import shouldn't be needed

mtrezza replied:
removed

mtrezza commented on spec/test.spec.js:
```diff
@@ -514,102 +551,120 @@ describe('S3Adapter tests', () => {
-      calls.forEach((call) => {
-        expect(call.args[0].ACL).toBe('public-read');
-      });
+      const commandArg = s3ClientMock.send.calls.mostRecent().args[0];
+      expect(commandArg.input.ACL).toBe('public-read');
```
This and other tests seems to not ensure anymore that there was only 1 call. Is that because the adapter has changed, or did that just get lost in the migration?

vahidalizad replied:
Unfortunately, I don’t remember why that got lost 😅, and I’m not sure why we need this test again. But if you still think it's necessary, I can add it back in!

mtrezza replied:
I haven't looked into the purpose of these lines, but if they don't cause any issues adding them back, then I'd just add them back.

vahidalizad replied:
I added these checks and command instance checks as well

mtrezza commented:
The integration tests are [failing](https://github.com/parse-community/parse-server-s3-adapter/actions/runs/11463810030/job/31898539287?pr=221) for Parse Server 6. I assume that is expected, since the adapter will require Parse Server >=7.3.0 due to https://github.com/parse-community/parse-server/issues/9268, correct?

mtrezza commented on index.js:
```diff
@@ -203,30 +226,31 @@ class S3Adapter {
+    data.Body.on('end', () => res.end());
+    data.Body.on('error', e => {
+      res.status(404);
+      res.send(e.message);
```
Do we need a test for this?

vahidalizad replied:
Added this test as well.

mtrezza commented on index.js:
```diff
@@ -65,6 +83,16 @@ class S3Adapter {
+        secretAccessKey: options.secretKey,
+      };
+    } else if (options.credentials) {
+      s3Options.credentials = options.credentials;
```
Do we need a test for this?

vahidalizad replied:
Yes we need a test for this part and i added it.

mtrezza commented:
Thanks for updating this PR; I've left a few comment above; just some minor things

mman commented:
@mtrezza @vahidalizad Just wanted to confirm that I have finally updated my alpha staging environment with this PR and have not seen any issues in testing. Good job!

mtrezza commented:
Great, maybe someone could address the open review comments, then we can go ahead and merge.

vahidalizad commented:
Sorry for the delay, I’ve been quite busy with work recently. I really appreciate your review and feedback! I’ll address the comments over the coming weekend.

vahidalizad commented on index.js:
```diff
@@ -128,52 +154,50 @@ class S3Adapter {
+    await this.createBucket();
+    const command = new PutObjectCommand(params);
+    const response = await this._s3Client.send(command);
+    const location = `https://${this._bucket}.s3.${this._region}.amazonaws.com/${params.Key}`;
```
In Parse Server's FileRouter (line 237), the url and name are used right after file creation, so we need a valid URL immediately. Since the S3 SDK doesn’t provide a URL after creating the file, I manually compose one using the bucket and key.

That said, I could update this to better support external endpoints, making it more flexible.

vahidalizad replied:
I added this part, could you please review it?

mtrezza commented:
@AdrianCurtin  we won't make any commits to the Parse Server 6 LTS branch anymore.

mtrezza commented:
It would be great if someone could get this over the finish line and address the open review questions.

vahidalizad commented:
I believe it’s ready for a final review. Apologies for the delay in getting this done work has been particularly busy lately.

mtrezza commented on index.js:
```diff
@@ -55,6 +73,7 @@ class S3Adapter {
     this._presignedUrlExpires = parseInt(options.presignedUrlExpires, 10);
     this._encryption = options.ServerSideEncryption;
     this._generateKey = options.generateKey;
+    this._endpoint = options.s3overrides?.endpoint;
```
Ensure this is documented

mtrezza replied:
It is in README under DO config.

mtrezza commented on index.js:
```diff
@@ -73,29 +102,27 @@ class S3Adapter {
+      this._hasBucket = true;
+    } catch (error) {
+      if (error.name === 'BucketAlreadyOwnedByYou') { this._hasBucket = true; }
+      else { throw error; }
```
```suggestion
      else {
        throw error;
      }
```

mtrezza commented on index.js:
```diff
@@ -73,29 +102,27 @@ class S3Adapter {
-        });
-      });
+  async createBucket() {
+    if (this._hasBucket) { return; }
```
```suggestion
    if (this._hasBucket) {
      return;
    }
```

mtrezza commented on index.js:
```diff
@@ -128,52 +155,51 @@ class S3Adapter {
+    await this.createBucket();
+    const command = new GetObjectCommand(params);
+    const response = await this._s3Client.send(command);
+    if (response && !response.Body) { throw new Error(response); }
```
```suggestion
    if (response && !response.Body) {
      throw new Error(response);
    }
```

mtrezza commented on index.js:
```diff
@@ -203,30 +228,31 @@ class S3Adapter {
+    await this.createBucket();
+    const command = new GetObjectCommand(params);
+    const data = await this._s3Client.send(command);
+    if (data && !data.Body) { throw new Error('S3 object body is missing.'); }
```
```suggestion
    if (data && !data.Body) {
      throw new Error('S3 object body is missing.');
    }
```

mtrezza commented on spec/test.spec.js:
```diff
@@ -18,7 +20,7 @@ describe('S3Adapter tests', () => {
     expect(() => {
       new S3Adapter('accessKey', 'secretKey', {});
-    }).toThrow(new Error('Failed to configure S3Adapter. Arguments don\'t make sense'));
+    }).toThrow(new Error("Failed to configure S3Adapter. Arguments don't make sense"));
```
Revert
```suggestion
    }).toThrow(new Error('Failed to configure S3Adapter with invalid arguments.'));
```

mtrezza commented on spec/test.spec.js:
```diff
@@ -134,8 +163,8 @@ describe('S3Adapter tests', () => {
-    it('should accept endpoint as an override option in args', () => {
-      const otherEndpoint = new AWS.Endpoint('nyc3.digitaloceanspaces.com');
+    it('should accept endpoint as an override option in args', async () => {
+      const otherEndpoint = 'https://test.com:8080/path?foo=bar';
```
```suggestion
      const otherEndpoint = 'https://example.com:8080/path?foo=bar';
```

mtrezza commented on spec/test.spec.js:
```diff
@@ -144,20 +173,39 @@ describe('S3Adapter tests', () => {
+      expect(endpointFromConfig.protocol).toEqual('https:');
+      expect(endpointFromConfig.path).toEqual('/path');
+      expect(endpointFromConfig.port).toEqual(8080);
+      expect(endpointFromConfig.hostname).toEqual('test.com');
```
```suggestion
      expect(endpointFromConfig.hostname).toEqual('example.com');
```

mtrezza commented on spec/test.spec.js:
```diff
@@ -275,30 +359,38 @@ describe('S3Adapter tests', () => {
+    it('should get using the baseUrl', async () => {
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://example.com/files/foo/bar/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Should be `await expectAsync(s3.getFileLocation...).toBeResolvedWith(...);`

mtrezza commented on spec/test.spec.js:
```diff
@@ -275,30 +359,38 @@ describe('S3Adapter tests', () => {
       options.baseUrlDirect = true;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://example.com/files/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -275,30 +359,38 @@ describe('S3Adapter tests', () => {
       options.directAccess = false;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://my.server.com/parse/files/xxxx/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -275,30 +359,38 @@ describe('S3Adapter tests', () => {
       delete options.baseUrl;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('https://my-bucket.s3.amazonaws.com/foo/bar/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -318,30 +410,38 @@ describe('S3Adapter tests', () => {
+    it('should get using the baseUrl', async () => {
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://example.com/files/foo/bar/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -318,30 +410,38 @@ describe('S3Adapter tests', () => {
       options.baseUrlDirect = true;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://example.com/files/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -318,30 +410,38 @@ describe('S3Adapter tests', () => {
       options.directAccess = false;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://my.server.com/parse/files/xxxx/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -318,30 +410,38 @@ describe('S3Adapter tests', () => {
       delete options.baseUrl;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('https://my-bucket.s3.amazonaws.com/foo/bar/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -357,67 +457,78 @@ describe('S3Adapter tests', () => {
+    it('should get using the baseUrl', async () => {
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://example.com/files/foo/bar/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -357,67 +457,78 @@ describe('S3Adapter tests', () => {
     });
 
-    it('when use presigned URL should use S3 \'getObject\' operation', () => {
+    it("when use presigned URL should use S3 'getObject' operation", async () => {
```
```suggestion
    it('when use presigned URL should use S3 \'getObject\' operation', async () => {
```

mtrezza commented on spec/test.spec.js:
```diff
@@ -357,67 +457,78 @@ describe('S3Adapter tests', () => {
       options.baseUrlDirect = true;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://example.com/files/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -357,67 +457,78 @@ describe('S3Adapter tests', () => {
       options.directAccess = false;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('http://my.server.com/parse/files/xxxx/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -357,67 +457,78 @@ describe('S3Adapter tests', () => {
       delete options.baseUrl;
       const s3 = new S3Adapter('accessKey', 'secretKey', 'my-bucket', options);
-      expect(s3.getFileLocation(testConfig, 'test.png')).toEqual('https://my-bucket.s3.amazonaws.com/foo/bar/test.png');
+      expect(await s3.getFileLocation(testConfig, 'test.png')).toEqual(
```
Use `expectAsync`

mtrezza commented on spec/test.spec.js:
```diff
@@ -514,102 +625,144 @@ describe('S3Adapter tests', () => {
       const s3 = getMockS3Adapter(options);
-      spyOn(s3._s3Client, 'upload').and.callThrough();
-      // Save file
+      s3._s3Client = s3ClientMock;
```
Why are we using a mock client instead of spying? The issue with this is that we are now testing a mock client, which may not behave like the actual client if that changes, so the test becomes less robust.

vahidalizad replied:
The issue here is not with the client itself, as I’ve consistently defined `const s3ClientMock = jasmine.createSpyObj('S3Client', ['send']);` and spied on the client to ensure realistic behavior.

What I believe you’re referring to is the handling of the adapter itself. I’ve been mindful of this and have aimed to minimize the use of mocks for the adapter, sticking to the original adapter whenever possible. Specifically, I’ve kept the mocked adapter only in tests that are focused exclusively on the logic within the adapter itself and not on the client.

vahidalizad replied:
If I misunderstood your point and you’re referring to why we’re using a mock client instead of testing with the actual S3 client, it’s primarily because we don’t have the necessary environment variables set up to enable the client’s realistic behavior during testing.

If you believe having realistic behavior is essential, we could set up a fake S3 (like using LocalStack or a similar tool) and test against it. This would allow us to avoid defining mock behaviors and instead validate interactions with a closer-to-real implementation of the S3 client.

mtrezza replied:
I think it's fine as is, thanks for the details.

mtrezza commented:
Great work, just a few tiny things that are quickly to resolve. One thing however is that instead of spying in some tests it is using a mock S3 client, which we should avoid if possible. Could you revert that back to spying?

mtrezza reviewed (approved):
Looks good! Thanks for your effort. I think this is good to merge. @mman have you found anything unexpected while you've been using the adapter?

jimnor0xF commented:
Tried running this in my env, ran into some issues since I am using the `s3overrides` config option and passing credentials explicitly at the moment.

Erroring out with: `error: Could not load credentials from any providers`

```javascript
import S3Adapter from '@parse/s3-files-adapter';

const awsS3Config = {
  bucket: R2_BUCKET,
  region: 'auto',
  directAccess: true,
  fileAcl: 'none',
  baseUrl: R2_BASE_URL,
  baseUrlDirect: true,
  signatureVersion: 'v4',
  s3overrides: {
    accessKeyId: R2_ACCESS_KEY_ID,
    secretAccessKey: R2_ACCESS_KEY_SECRET,
    endpoint: `https://${CF_ACCOUNT_NUMBER}.r2.cloudflarestorage.com`,
  },
};


export const config = {
...
  filesAdapter: new S3Adapter(awsS3Config),
}

```

However, I see that in README.md the following is mentioned, so this should be working still?

>If for some reason you really need to be able to set the key and secret explicitly, you can still do it using s3overrides as described below and setting accessKeyId and secretAccessKey in the s3Overrides object.```

jimnor0xF commented:
> However, I see that in README.md the following is mentioned, so this should be working still?
> 
> > If for some reason you really need to be able to set the key and secret explicitly, you can still do it using s3overrides as described below and setting accessKeyId and secretAccessKey in the s3Overrides object.

@vahidalizad 
Looks like this is because it was previously enough to [merge options.s3overrides](https://github.com/parse-community/parse-server-s3-adapter/blob/master/index.js#L74) into s3Options. 

I see that in v3 you now need to put the secret key/id under s3Options.credentials. That case is handled when you have [secretKey and accessKey right under options](https://github.com/parse-community/parse-server-s3-adapter/blob/1f74921cea769e5a703d7c95c629c46b46021103/index.js#L74). But not in the case when the secrets are put into `s3overrides`.


Besides above, I tested with a working configuration and it seems to be working perfectly. Thanks for your efforts! :smile:

mtrezza commented:
There should be no breaking changes, if they can be avoided. So it seems we're currently missing a test case from what @jimnor0xF found, because the CI passes. @vahidalizad could you please add a test for setting the credentials as they are set in v2, and adapt this PR to make the same test pass with v3?

mtrezza commented:
In general we want to avoid breaking changes if possible. It seems we're currently missing a test case from what @jimnor0xF found, because the CI passes. If the change is worth the breaking change, then we can adopt it and release as a new major version. Otherwise, it seems we're currently missing a test for how @jimnor0xF used to set the credentials in v2 and we'd adapt this PR to make the same test passes with v3.

vahidalizad commented:
Thank you for the feedback!

You're correct, the AWS SDK v3 moved all credential-related values from the main parameters into a dedicated credentials key. I believe this change was made to consolidate the growing number of credential-related parameters into a single, structured object for better organization.

In my opinion, it would be better to align with the SDK v3 approach to avoid potential confusion and ensure consistency with their design. We could document this change clearly in the README

That said, if you’d prefer to avoid the breaking change, I can adjust the PR to extend the main parameters for backwards compatibility. Should we limit this extension to just `accessKeyId` and `secretAccessKey`, or include all credential-related fields?

Let me know your thoughts

mtrezza commented:
> In my opinion, it would be better to align with the SDK v3 approach to avoid potential confusion and ensure consistency with their design. We could document this change clearly in the README

I'd agree. Could you add an entry to the README? We'll then publish a release 4.0 with your PR.

mman commented:
@vahidalizad Could you please take a look at my issue and suggested fixes at https://github.com/parse-community/parse-server-s3-adapter/issues/230. If I am not mistaken in my investigation the AWS S3 JS SDK v3 changed the behavior from of `CreateBucketCommand` compared to v2 and breaks Digital Ocean Spaces compatibility.

vahidalizad commented:
@mtrezza I’ve added the breaking change details to the README, merged the fixes suggested by @mman into the adapter, and included bucket creation tests and credential tests. Could you review and confirm if we’re ready to merge?

mtrezza commented on package.json:
```diff
@@ -38,6 +39,7 @@
     "nyc": "17.1.0",
     "parse": "3.3.1",
     "parse-server": "7.3.0",
+    "rewire": "7.0.0",
```
Why do we need this new dependency? Is there a Node built-in alternative?

vahidalizad replied:
I needed to unit test the S3 adapter's construction, particularly to verify the credentials being passed to the `S3Client` from the adapter during initialization.
Instead of creating a mock version of the adapter (and testing the mock instead of the original), I chose to directly test the real adapter. However, I needed to spy on the S3Client, which is imported from the `@aws-sdk/client-s3` library inside the adapter file, to inspect these inputs. Since S3Client is internal to the module, I used rewire to access and turn it into a spy, enabling me to validate the credentials and other parameters being passed without modifying the production code.

vahidalizad replied:
There isn’t a Node.js built-in alternative that provides the same capability as rewire for accessing and modifying private or internal variables within a module.
If avoiding rewire is a priority, an alternative approach would be to refactor the adapter code, exposing the S3Client as a dependency via dependency injection.
That said, I believe going forward with rewire doesn’t raise concerns because it is added as a dev dependency strictly for testing purposes and does not impact or modify the production code.

mtrezza replied:
I believe we can keep it

mtrezza commented:
What is the state of this PR, has this been tested thoroughly to get it ready for merge?

vahidalizad commented:
> What is the state of this PR, has this been tested thoroughly enough to get it ready for merge?

I tested this with my local S3 setup, but I haven’t done any further testing on AWS or DigitalOcean. Maybe @mman has done additional testing?

mman commented:
> What is the state of this PR, has this been tested thoroughly enough to get it ready for merge?

I have tested early versions of this PR against Amazon S3 successfully, as well as recent version against Digital Ocean Spaces using FullAccess key, as well as Read/Write/Delete key. I'd put it into prod as soon as it is tagged...

Martin

mtrezza commented:
@vahidalizad Just a quick ping here, would you available for a quick chat on Slack?

vahidalizad commented:
> @vahidalizad Just a quick ping here, would you available for a quick chat on Slack?

Hey @mtrezza , sure! Just ping me on Slack whenever you're ready. You can invite me at azizalizad@gmail.com.
